### PR TITLE
fix ternary op in subset of std algorithms not working with nvhpc

### PR DESCRIPTION
**Pull request by fnrizzi**
fb8179f4bae685e8fc29c9fdd890b41e4c8b92ff forced skipping a subset of the std algorithms tests because failing for the NVHPC CI build. 
Looks like the problem was the use of a ternary operator which hvc++ (version 22.9) did not like for some reasons, so using a simpler `if` fixes it and all tests pass. 

NOTE: this is related to but should be merged before #5150 

Co-authored-by: Cezary Skrzyński

### CODE REVIEW

**Review by dalg24 — COMMENTED** [review]
Please improve the description.

File: algorithms/src/std_algorithms/impl/Kokkos_SearchN.hpp [hunk 1]
```diff
@@ -59,9 +59,16 @@ struct StdSearchNFunctor {
...
+    /* FRIZZI: 05/2023
+       Originally the code below was using a ternary operator but nvc++ for 22.9
+       did not work with that, which was the reason for
+       fb8179f4bae685e8fc29c9fdd890b41e4c8b92ff Using the "simpler" code below
+       works.
+    */
+    red_value_type rv = {::Kokkos::reduction_identity<IndexType>::min()};
+    if (found) {
+      rv.min_loc_true = i;
+    }
```
[the remark this answers is not in the document]
Comment (fnrizzi) [reply algorithms/src/std_algorithms/impl/Kokkos_SearchN.hpp:71]: yes, i thought about guarding that way, but is that worth it? I feel like the simpler code is all we need, why keeping all that extra ?

[the remark this answers is not in the document]
Comment (dalg24) [reply algorithms/src/std_algorithms/impl/Kokkos_SearchN.hpp:71]: I tend to think that is something that the NVHPC compiler team needs to fix and that way this is easily greppable.
Also I thought the comment was lengthy and that the code suggested expressed what I would have like to see instead.
If others agree with you I can be convinced.  I would still want the comment to be shortened in that case.

[the remark this answers is not in the document]
Comment (fnrizzi) [reply algorithms/src/std_algorithms/impl/Kokkos_SearchN.hpp:71]: yes i am happy to shorten the message, but i think it is important to keep it there in some form in all instances

File: algorithms/src/std_algorithms/impl/Kokkos_Mismatch.hpp [hunk 2]
```diff
@@ -27,9 +27,10 @@ namespace Kokkos {
 namespace Experimental {
 namespace Impl {
 
-template <class IndexType, class IteratorType1, class IteratorType2,
-          class ReducerType, class BinaryPredicateType>
+template <class IteratorType1, class IteratorType2, class ReducerType,
+          class BinaryPredicateType>
 struct StdMismatchRedFunctor {
+  using index_type     = typename IteratorType1::difference_type;
```
[the remark this answers is not in the document]
Comment (fnrizzi) [reply algorithms/src/std_algorithms/impl/Kokkos_Mismatch.hpp:33]: yes i think it was and slipped, i reverted back to making the functor templated on the index type too as before

File: algorithms/src/std_algorithms/impl/Kokkos_SearchN.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (fnrizzi) [reply algorithms/src/std_algorithms/impl/Kokkos_SearchN.hpp]: changed to what Daniel suggested, and then we will see during meeting if we want to keep the "simpler" code or not

File: algorithms/src/std_algorithms/impl/Kokkos_IsSortedUntil.hpp [hunk 3]
```diff
@@ -28,33 +28,30 @@ namespace Kokkos {
...
-  StdIsSortedUntilFunctor(IteratorType _first1, IndicatorViewType indicator,
-                          ComparatorType comparator)
-      : m_first(std::move(_first1)),
-        m_indicator(std::move(indicator)),
-        m_comparator(std::move(comparator)) {}
+  StdIsSortedUntilFunctor(IteratorType first, ComparatorType comparator,
+                          ReducerType reducer)
+      : m_first(std::move(first)),
+        m_comparator(std::move(comparator)),
+        m_reducer(std::move(reducer)) {}
```
Comment (dalg24) [inline algorithms/src/std_algorithms/impl/Kokkos_IsSortedUntil.hpp:54]: What about the changes to that file?
These are not related to the issue with the ternary conditional operator.

Comment (fnrizzi) [reply algorithms/src/std_algorithms/impl/Kokkos_IsSortedUntil.hpp:54]: was editing the description while you were adding your comment. 
`is_sorted_until` was not working but did not have a ternary so we changed the impl to use a reduce and avoid an extra allocation that was previously used in the scan.

**Comment by fnrizzi** [comment]
@dalg24 
two options: 
1) i can try to revert is_sorted_until to what it was before and see if it works so that i can make this PR just about the ternary for nvhpc
2) i can leave the PR as is since the title is making algorithms work for nvhpc so it covers moe broadly all std failing for nvhpc

File: algorithms/src/std_algorithms/impl/Kokkos_IsSortedUntil.hpp (hunk 3 above)
Comment (fnrizzi) [reply algorithms/src/std_algorithms/impl/Kokkos_IsSortedUntil.hpp:54]: the reason is_sorted_until was not working before was becuase it was using find inside which had a ternary op

File: algorithms/src/std_algorithms/impl/Kokkos_IsSortedUntil.hpp [hunk 4]
```diff
@@ -73,40 +70,31 @@ IteratorType is_sorted_until_impl(const std::string& label,
...
+      // use CTAD
+      StdIsSortedUntilFunctor(first, comp, reducer), reducer);
+
+  /* If the reduction result is equal to the initial value,
+     and it means the range is sorted until the end */
+  if (red_result == red_result_init) {
+    return last;
+  } else {
+    /* If  such index is found, then the range is sorted until there and
+       we need to return an iterator past the element found so do +1 */
```
Comment (masterleinad) [inline algorithms/src/std_algorithms/impl/Kokkos_IsSortedUntil.hpp:95]:
```suggestion
    /* If such an index is found, then the range is sorted until there and
       we need to return an iterator past the element found so do +1 */
```

**Review by masterleinad — APPROVED** [review]
Looks good to me. It's much better to use `parallel_reduce` than `parallel_scan` if we can.

**Comment by fnrizzi** [comment]
> It's much better to use `parallel_reduce` than `parallel_scan` if we can.

agreed!

File: algorithms/src/std_algorithms/impl/Kokkos_IsSortedUntil.hpp (hunk 4 above)
Comment (fnrizzi) [reply algorithms/src/std_algorithms/impl/Kokkos_IsSortedUntil.hpp]: done

**Comment by dalg24** [comment]
> @dalg24 two options:
> 
> 1. i can make this PR just about the ternary for nvhpc and then post another one to improve is_sorted_until
> 2. or i can leave the PR as is since the title is already broadly referrring to "making algorithms work for nvhpc" so all changes here fit anyway
> 
> whatever you prefer

I prefer making this PR just about the ternary conditional operator.

**Review by dalg24 — APPROVED** [review]

**Comment by dalg24** [comment]
Ignoring HIP builds that timed out
